File: parser.js
import { Sender, Request } from "./types.js";
import * as fs from "fs";

/**
 * Takes in a data file and returns two lists, one containing all the Senders and
 the other containing all the Requests.
 * @param {string} fileName Name of log file to process
 * @returns {senders: [Senders], requests: [Requests]}
 */
 export function processDataFile(fileName = "./test.log") {
  const senders = [];
  const requests = [];
  try {
    const data = fs.readFileSync(fileName, "utf8");
    const lines = data.split("\n");
    lines.forEach((line) => {
      const tokens = line.split('"');
      if (tokens.length > 2) {
        senders.push(getSenderData(tokens[0]));
        requests.push(getRequestData(tokens[1]));
      }
    });

    return { senders, requests };
  } catch (error) {
    return error
  }
}

/**
 * Takes in the string from the log entry containing the sender data and returns
 * an object containing the details of the sender
 * @param {string} senderData String from log containing sender data
 * @returns Sender
 */
export function getSenderData(senderData) {
  const senderDataList = senderData.split(" ");
  return new Sender(
    senderDataList[0],
    senderDataList[3].slice(1, 12),
    senderDataList[3].slice(13, 21)
  );
}

/**
 * Takes in string from log entry containing request data and returns Request object
 * @param {string} requestData String from log containing request data
 * @returns Request
 */
export function getRequestData(requestData) {
  const requestDataList = requestData.split(" ");
  return new Request(...requestDataList);
}

/**
 * Finds the most frequent IP in an array of senders
 * @param {[Sender]} senders Array of senders
 * @param {number} numCount Number of senders to return
 * @returns [string]
 */
export function countMostCommonIps(senders, numCount = 0) {
  if (numCount < 0) {
    return [];
  }
  const ipCounts = [];
  senders.forEach((sender) => {
    const ipIndex = ipCounts.findIndex((ipCount) => ipCount.ip == sender.ip);
    if (ipIndex != -1) {
      ipCounts[ipIndex].count += 1;
    } else {
      ipCounts.push({ ip: sender.ip, count: 1 });
    }
  });
  ipCounts.sort((a, b) => b.count - a.count);
  if (numCount) {
    return ipCounts.slice(0, numCount).map((ipCount) => ipCount.ip);
  } else {
    return ipCounts.map((ipCount) => ipCount.ip);
  }
}

/**
 * Finds the most frequent URL in an array of requests
 * @param {Request} requests Array of requests
 * @param {number} numCount Number of URLs to return
 * @returns [string]
 */
export function countMostCommonUrls(requests, numCount = 0) {
  if (numCount < 0) {
    return [];
  }
  const urlCounts = [];
  requests.forEach((request) => {
    const urlIndex = urlCounts.findIndex(
      (urlCount) => urlCount.url == request.url
    );
    if (urlIndex != -1) {
      urlCounts[urlIndex].count += 1;
    } else {
      urlCounts.push({ url: request.url, count: 1 });
    }
  });

  urlCounts.sort((a, b) => b.count - a.count);
  if (numCount) {
    return urlCounts.slice(0, numCount).map((urlCount) => urlCount.url);
  } else {
    return urlCounts.map((urlCount) => urlCount.url);
  }
}

export function printResults(
  ipCounts,
  urlCounts,
  mostCommonIps,
  mostCommonUrls
) {}

File: parser.py
import re
from collections import defaultdict

def getSenderData(senderData):
  senderDataDict = {}
  senderDataList = senderData.split(" ")
  # Assumption: the length of the IP, date, and time encodings in the log file
  # will be consistent
  senderDataDict["ip"] = senderDataList[0]
  senderDataDict["date"] = senderDataList[3][1:12]
  senderDataDict["time"] = senderDataList[3][13:] + " " + senderDataList[4][:-1]
  return senderDataDict
  
def getRequestData(requestData):
  requestDataList = requestData.split(' ')
  return {
    'requestType': requestDataList[0],
    'url': requestDataList[1],
    'protocol': requestDataList[2]
  }
  
def getStatusData(statusData):
  pass
def getMachineData(machineData):
  pass

# Given a dictionary of IP counts and URL counts, returns the 3 most common IPs
# and URLs to the caller
def countMostCommon(ipCounts, urlCounts):
  mostCommonIps = sorted(ipCounts, key=ipCounts.get, reverse=True)[:3]
  mostCommonUrls = sorted(urlCounts, key=urlCounts.get, reverse=True)[:3]
  
  return (mostCommonIps, mostCommonUrls)

def printResults(ipCounts, urlCounts, mostCommonIps, mostCommonUrls):
  print("Number of unique IP addresses: ", len(ipCounts.keys()))
  print("Top 3 most common IP addresses: ")
  for ip in mostCommonIps:
    print(ip, ipCounts[ip])
  print("Top 3 most visited URLs:")
  for url in mostCommonUrls:
    print(url, urlCounts[url])

def processDataFile(fileName='./test.log'):
  file = open(fileName)
  ipCounts = defaultdict(int)
  urlCounts = defaultdict(int)
  for line in file:
    tokens = line.split("\"")
    senderData = getSenderData(tokens[0])
    requestData = getRequestData(tokens[1])
    ipCounts[senderData["ip"]] += 1
    urlCounts[requestData["url"]] += 1
  return (ipCounts, urlCounts)
  
  
    

if __name__ == "__main__":
  ipCounts, urlCounts = processDataFile()
  mostCommonIps, mostCommonUrls = countMostCommon(ipCounts, urlCounts)
  printResults(ipCounts, urlCounts, mostCommonIps, mostCommonUrls)
    
    



